engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java

```diff
@@ -191,6 +199,51 @@ Text toDisplayString(boolean allowSideEffects, @CachedLibrary("this") InteropLib
     }
   }
 
+  @ExportMessage
+  int identityHashCode(
+      @CachedLibrary(limit = "5") InteropLibrary fieldsInterop,
+      @Cached(value = "computeIdentityHashCode(this, fieldsInterop)", allowUncached = true) int cachedHashCode
+  ) {
+    return cachedHashCode;
+  }
+
+  @ExportMessage
+  static class IsIdenticalOrUndefined {
+    @Specialization
+    static TriState isIdenticalOrUndefined(
+        Atom thisAtom,
+        Atom otherAtom,
+        @CachedLibrary(limit = "5") InteropLibrary fieldsInterop,
+        @Cached(value = "computeIdentityHashCode(thisAtom, fieldsInterop)", allowUncached = true) int thisHashCode,
+        @Cached(value = "computeIdentityHashCode(otherAtom, fieldsInterop)", allowUncached = true) int otherHashCode
+    ) {
+      return thisHashCode == otherHashCode ? TriState.TRUE : TriState.FALSE;
+    }
+
+    @Fallback
+    static TriState fallBack(Atom thisAtom, Object other) {
+      return TriState.FALSE;
+    }
+  }
+
+  @ExplodeLoop
+  static int computeIdentityHashCode(Atom atom, InteropLibrary fieldsInterop) {
+    int[] hashCodes = new int[atom.fields.length];
+    for (int i = 0; i < atom.fields.length; i++) {
+      Object field = atom.fields[i];
+      if (fieldsInterop.hasIdentity(field)) {
+        try {
+          hashCodes[i] = fieldsInterop.identityHashCode(field);
+        } catch (UnsupportedMessageException e) {
+          throw new PanicException(atom, fieldsInterop);
+        }
+      } else {
+        hashCodes[i] = field.hashCode();
```

Comment:
I am afraid this is not going to PE well. You'd probably have to use _class condition profile_ to make sure the type of the `field` is always the same.

Reply:
Probably yes. And also a `ConditionProfile` for `if (fieldsInterop.hasIdentity(field))`. Anyway, the most important thing at this moment here is the gist of how the hash code is computed for an atom by default. It is consistent with `Any.==` - it recurses on the fields of the Atom.